### PR TITLE
Add template janitor sanity check / enhancements

### PR DESCRIPTION
**Pull request by lefticus**

### CODE REVIEW

**Comment by lefticus** [comment]
Adding remaining tests for #3

**Comment by LtdSauce** [comment]
@lefticus not sure if you are working on the feature to turn a project into a template using the github cli. If so i wanted to make you aware that i see a potential race condition between the janitor and the creator.

Either both have to be aware of the special meaning of the suffix or the janitor has to call the other action after turning the project into a template.

**Comment by lefticus** [comment]
@LtdSauce I wasn't going to do the `_template` feature right now (not sure if it makes sense or not, honestly)

What I have attempted (and seems to be working)

 * Test to make sure project is still configurable
 * automatic update of template-names in the case that the project is a template, but the name of the project has changed.
 
That is:

* You can fork, and it will change the names in the main README to match the names of the new template-name
* you can "use template" and it will do everything it was already doing

This should help with maintenance when creating new forks

**Comment by lefticus** [comment]
@LtdSauce I think this is ready now, your review would be greatly appreciated.

**Comment by LtdSauce** [comment]
> @LtdSauce I wasn't going to do the `_template` feature right now (not sure if it makes sense or not, honestly)

I guess forking and renaming it afterwards is fine. The suffix "hack" would make it a little bit easier to fork as the renaming step is directly done.

I'll try to have a look at your changes tomorrow.

**Comment by lefticus** [comment]
> I guess forking and then renaming it afterwards is fine. The suffix "hack" would make it a little bit easier to fork as the renaming step is directly done.
> 

With what I did, it should track the name changes whenever they happen. So it will track the forked owner name, then if you rename the fork, will track the new project name... if I did it right.

**Comment by LtdSauce** [comment]
> 
> > I guess forking and then renaming it afterwards is fine. The suffix "hack" would make it a little bit easier to fork as the renaming step is directly done.
> > 
> 
> With what I did, it should track the name changes whenever they happen. So it will track the forked owner name, then if you rename the fork, will track the new project name... if I did it right.
> 

Hmm... tracking the project name seems reasonable. But i see one caveat with the owner. Contributing to the template by forking and then opening a PR will always have an unwanted diff because the org/owner will be replaced in every fork. Maybe the renaming should only happen when the project name changes?

**Comment by lefticus** [comment]
> 
> Hmm... tracking the project name seems reasonable. But i see one caveat with the owner. Contributing to the template by forking and then opening a PR will always have an unwanted diff because the org/owner will be replaced in every fork. Maybe the renaming should only happen when the project name changes?

That's a valid point ... but in a normal fork the badges will always be broken otherwise.

Hmm..

The check to commit is here, so it's easy to adjust for: https://github.com/cpp-best-practices/cpp_boilerplate_project/blob/template-janitor-testing/.github/workflows/template-creator.yml#L78

**Comment by LtdSauce** [comment]
On the other side... After the PR gets merged the create workflow will fix the badges again...

But i think it is okay for a contribution fork to have the badges point to the root repo. At least i only look at the github actions directly triggered by my commits. Futhermore, the forked project has to be created at codecov and lgtm seperately. Which i do not think a regular contributor will do, don't you think?

File: .github/workflows/template-creator.yml
Comment (LtdSauce) [inline .github/workflows/template-creator.yml:20]: nit: i think it is more like a `template-rename` than a create. (Same goes for the file name)

File: .github/workflows/template-creator.yml
Comment (LtdSauce) [inline .github/workflows/template-creator.yml:31]: `NEW_URL` is unused in this action

File: .github/workflows/template-creator.yml
Comment (LtdSauce) [inline .github/workflows/template-creator.yml:49]: nit: `NEW_PROJECT` could be used.

File: .github/workflows/template-creator.yml
Comment (LtdSauce) [inline .github/workflows/template-creator.yml:77]: There is something missing
```suggestion
        # only commit and push if we are a template and project name has changed
```

File: .github/workflows/template-creator.yml
Comment (LtdSauce) [inline .github/workflows/template-creator.yml:78]: I'd use the name as a trigger to commit.
```suggestion
        if: fromJson(steps.get_repo_meta.outputs.data).is_template == true && env.TEMPLATE_NAME != env.NEW_PROJECT
```

**Review by LtdSauce — CHANGES_REQUESTED** [review]
As i already mentioned i think only applying the changes when the `TEMPLATE_NAME` has changed seems like the more general case.

Also some little nits with the variables and the action name. Actually i also wasn't consistent with using my `NEW_` variables, i just noticed.

**Comment by lefticus** [comment]
@LtdSauce I believe I have addressed all of your comments.

**Comment by LtdSauce** [comment]
> @LtdSauce I believe I have addressed all of your comments.

@lefticus you did. But my review was incomplete :/

**Comment by LtdSauce** [comment]
I meant i missed something. I finished my review. And now i think everything is covered :)

**Review by LtdSauce — CHANGES_REQUESTED** [review]
Sorry i missed this one in my review!

File: .github/workflows/template-janitor.yml
Comment (LtdSauce) [reply .github/workflows/template-janitor.yml:76]: and maybe it it better to omit the  `-rf` option. As those will hide the current problem. I think it is better if this fails if the file is missing

File: .github/workflows/template-janitor.yml
Comment (LtdSauce) [inline .github/workflows/template-janitor.yml:75]:
```suggestion
          rm .github/workflows/template-janitor.yml
```
I think this one should fail if the file is missing to avoid forgetting this if the janitor is beeing renamed

File: .github/workflows/template-janitor.yml
Comment (LtdSauce) [inline .github/workflows/template-janitor.yml:71]: now.... when i think about it, maybe this one should omit the `-f` option. I think it would be nice if this one fails when the `TEMPLATES_PATH` changes without the proper change to the janitor.

Just noticed that the janitor action did not fail although the creator was renamed to renamer... that got me thinking
```suggestion
          rm -r ${{ env.TEMPLATES_PATH }}
```

**Review by LtdSauce — CHANGES_REQUESTED** [review]
Sorry that my review is split up.... but now i think i realy have everything!

File: .github/workflows/template-janitor.yml
Comment (LtdSauce) [inline .github/workflows/template-janitor.yml:49]:
```suggestion
          sed -i "s|%%myurl%%|${{ env.NEW_URL }}|gi" CMakeLists.txt
```
Somehow this ended up empty in the last run

**Review by LtdSauce — CHANGES_REQUESTED** [review]
@lefticus now that i already have many reviews created... this one should not hurt too ... -.- sorry!

Now i am realy realy finished...

File: .github/workflows/template-janitor.yml
Comment (LtdSauce) [inline .github/workflows/template-janitor.yml:49]:
```suggestion
          sed -i "s|%%myurl%%|${{ fromJson(steps.get_repo_meta.outputs.data).html_url }}|gi" CMakeLists.txt
```

Just noticed that the repositoryUrl from the env contains the git link (git://github.com/cpp-best-practices/cpp_boilerplate_project.git). See action runs.

maybe this one is what realy needs to go here? See https://docs.github.com/en/rest/reference/repos#get-a-repository

**Review by LtdSauce — CHANGES_REQUESTED** [review]
Huh.... my "fix" for the URL is also broken. This should work though...

The other changes seem fine.

File: .github/workflows/template-janitor.yml
Comment (lefticus) [reply .github/workflows/template-janitor.yml:49]: haha, thanks for the back and forth here.

There really are very many different moving pieces, and things that are difficult to test when it comes to CMake.

Example: the `option`s were recently broken and it went unnoticed for a time.

File: .github/workflows/template-renamer.yml
Comment (LtdSauce) [reply .github/workflows/template-renamer.yml:44]: Not sure if you noticed this this comment. So: bump!

Comment (LtdSauce) [reply .github/workflows/template-renamer.yml:44]: I just tested this. It is somewhat broken (notice that the lgtm links are not completely changed):
```
-[![ci]([https://github.com/cpp-best-practices/cpp_boilerplate_project/actions/workflows/ci.yml/badge.svg)](https://github.com/cpp-best-practices/cpp_boilerplate_project/actions/workflows/ci.yml](https://github.com/cpp-best-practices/cpp_boilerplate_project/actions/workflows/ci.yml/badge.svg](https://github.com/cpp-best-practices/cpp_boilerplate_project/actions/workflows/ci.yml)))
-[![codecov]([https://codecov.io/gh/cpp-best-practices/cpp_boilerplate_project/branch/main/graph/badge.svg)](https://codecov.io/gh/cpp-best-practices/cpp_boilerplate_project](https://codecov.io/gh/cpp-best-practices/cpp_boilerplate_project/branch/main/graph/badge.svg](https://codecov.io/gh/cpp-best-practices/cpp_boilerplate_project)))
-[![Language grade: C++]([https://img.shields.io/lgtm/grade/cpp/github/cpp-best-practices/cpp_boilerplate_project)](https://lgtm.com/projects/g/cpp-best-practices/cpp_starter_project/context:cpp](https://img.shields.io/lgtm/grade/cpp/github/cpp-best-practices/cpp_boilerplate_project](https://lgtm.com/projects/g/cpp-best-practices/cpp_starter_project/context:cpp)))
-[![CodeQL]([https://github.com/cpp-best-practices/cpp_boilerplate_project/actions/workflows/codeql-analysis.yml/badge.svg)](https://github.com/cpp-best-practices/cpp_boilerplate_project/actions/workflows/codeql-analysis.yml](https://github.com/cpp-best-practices/cpp_boilerplate_project/actions/workflows/codeql-analysis.yml/badge.svg](https://github.com/cpp-best-practices/cpp_boilerplate_project/actions/workflows/codeql-analysis.yml)))
+[![ci]([https://github.com/LtdSauce/test_template_renamer/actions/workflows/ci.yml/badge.svg)](https://github.com/LtdSauce/test_template_renamer/actions/workflows/ci.yml](https://github.com/LtdSauce/test_template_renamer/actions/workflows/ci.yml/badge.svg](https://github.com/LtdSauce/test_template_renamer/actions/workflows/ci.yml)))
+[![codecov]([https://codecov.io/gh/LtdSauce/test_template_renamer/branch/main/graph/badge.svg)](https://codecov.io/gh/LtdSauce/test_template_renamer](https://codecov.io/gh/LtdSauce/test_template_renamer/branch/main/graph/badge.svg](https://codecov.io/gh/LtdSauce/test_template_renamer)))
+[![Language grade: C++]([https://img.shields.io/lgtm/grade/cpp/github/LtdSauce/test_template_renamer)](https://lgtm.com/projects/g/cpp-best-practices/cpp_starter_project/context:cpp](https://img.shields.io/lgtm/grade/cpp/github/LtdSauce/test_template_renamer](https://lgtm.com/projects/g/cpp-best-practices/cpp_starter_project/context:cpp)))
+[![CodeQL]([https://github.com/LtdSauce/test_template_renamer/actions/workflows/codeql-analysis.yml/badge.svg)](https://github.com/LtdSauce/test_template_renamer/actions/workflows/codeql-analysis.yml](https://github.com/LtdSauce/test_template_renamer/actions/workflows/codeql-analysis.yml/badge.svg](https://github.com/LtdSauce/test_template_renamer/actions/workflows/codeql-analysis.yml)))
```
See https://github.com/LtdSauce/test_template_renamer/runs/5459863422?check_suite_focus=true

**Review by LtdSauce — CHANGES_REQUESTED** [review]
@lefticus except the broken lgtm link in the renamer i think this PR is ready.
I put a patch that is a possible fix in my comment below.

**Comment by lefticus** [comment]
@LtdSauce I had a busy week last week and will be digging into this today.

**Comment by lefticus** [comment]
I just noticed that the bug is because of an actual bug in the README.md file. I think fixing the README will fix the underlying issue, and I'll start with that.

I avoided the suggestion you make because it breaks other things I wanted to preserve in the readme, such as the link over to `cpp_starter_project`

**Comment by lefticus** [comment]
@LtdSauce I added automated testing with test names (which I kind of regret now, because it took so long) and tested it all here: https://github.com/ColoradoPlusPlus/test_boilerplate_project/pull/1/commits/0b34d3b9559e4adeee209b04b6a21fc3cafdf4db

It appears to be working in all paths now.

File: .github/workflows/template-renamer.yml
Comment (LtdSauce) [reply .github/workflows/template-renamer.yml:53]: Maybe it is less confusing to remove them?
```suggestion
```

**Review by LtdSauce — APPROVED** [review]
> I just noticed that the bug is because of an actual bug in the README.md file. I think fixing the README will fix the underlying issue, and I'll start with that.

Wow... sorry i missed that one in my review 🤦‍♂️ 

I think it looks good now. One minor optional change request. But i don't mind if you ignore it 🤷

File: .github/workflows/template-renamer.yml
Comment (lefticus) [reply .github/workflows/template-renamer.yml:53]: Right, that is a copy-and-paste unnecessary propagation.

**Review by LtdSauce — APPROVED** [review]

**Comment by lefticus** [comment]
considering the size of this diff, it has to be the most vetted and discussed PR ever between just two people. Thanks @LtdSauce

**Pull request merged**
